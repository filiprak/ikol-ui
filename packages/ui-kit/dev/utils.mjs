import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
