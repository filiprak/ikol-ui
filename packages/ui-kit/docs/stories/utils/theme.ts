import { useTheme } from '@ui/composables';
import { createIkolUI } from '@ui/index';
import { App } from 'vue';
import { StoryContext, VueRenderer } from '@storybook/vue3';

export function setupTheme(context: any) {
    const theme = useTheme();

    theme.is_dark.value = context?.globals?.theme === 'dark';

    return theme;
}

export function setupApp(app: App, context?: StoryContext<VueRenderer>) {
    app.use(createIkolUI());
}
