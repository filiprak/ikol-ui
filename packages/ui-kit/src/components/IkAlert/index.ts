import IkAlert from './IkAlert.vue';
export interface IkAlert extends InstanceType<typeof IkAlert> { };
export { IkAlert };
