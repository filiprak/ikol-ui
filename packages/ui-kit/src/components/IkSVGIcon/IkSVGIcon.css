.ik-svg-icon {
    display: inline-flex !important;
    justify-content: center;
    align-items: center;
}

.ik-svg-icon--clickable {
    cursor: pointer;
}

.ik-svg-icon--clickable:hover {
    opacity: 0.75;
}

.ik-svg-icon--disabled {
    opacity: 0.6;
    cursor: default;
    pointer-events: none;
}

/* circle */
.ik-svg-icon--circle {
    /* padding: 4px 5px 4px 5px; */
    border-radius: 50%;
    aspect-ratio: 1;
    background-color: var(--surface-darken-color);
}

.ik-svg-icon--circle.ik-svg-icon--default {
    background-color: var(--surface-darken-color);
}

.ik-svg-icon--circle.ik-svg-icon--primary {
    background-color: var(--primary-color-o1);
}

.ik-svg-icon--circle.ik-svg-icon--success {
    background-color: var(--success-color-o1);
}

.ik-svg-icon--circle.ik-svg-icon--accent {
    background-color: var(--accent-color-o1);
}

.ik-svg-icon--circle.ik-svg-icon--error {
    background-color: var(--error-color-o1);
}

/* filled */
.ik-svg-icon--filled.ik-svg-icon--default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

.ik-svg-icon--filled.ik-svg-icon--primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-svg-icon--filled.ik-svg-icon--success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-svg-icon--filled.ik-svg-icon--accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-svg-icon--filled.ik-svg-icon--error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}
