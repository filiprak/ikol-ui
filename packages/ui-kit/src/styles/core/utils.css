/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
