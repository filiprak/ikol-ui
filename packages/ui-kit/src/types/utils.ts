import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
