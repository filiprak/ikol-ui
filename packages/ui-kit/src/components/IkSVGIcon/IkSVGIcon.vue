<script setup lang="ts">
import '@ui/styles';
import './IkSVGIcon.css';

import { computed, ref, useAttrs, h, type PropType, watch, onMounted } from 'vue';
import { useTheme } from '@ui/composables/theme';
import { clamp, getTextColorForBackground, toCssUnits } from '@ui/utils/helpers';
import { useRender } from '@ui/composables/render';
import { fetchSvgText } from './loader';
import type { IkUIDesignColor } from '@ui/types/utils';

const props = defineProps({
    src: {
        type: String,
    },
    size_px: {
        type: Number,
        default: 28,
    },
    inner_scale: {
        type: Number,
        default: 0.6,
    },
    disabled: {
        type: Boolean,
        default: false,
    },
    circle: {
        type: Boolean,
        default: false,
    },
    design: {
        type: String as PropType<`${IkUIDesignColor}`>,
        default: 'default',
    },
    variant: {
        type: String,
    },
    color: {
        type: String,
    },
});

const icon = ref<HTMLDivElement | null>(null);
const attrs = useAttrs();
const theme = useTheme();

const is_filled = computed(() => {
    return props.variant === 'filled';
});

const bg_color = computed(() => {
    if (props.color) {
        return is_filled.value ? ('#' + props.color) : ('#' + props.color + '1A');
    } else {
        return null;
    }
});

const stroke = computed(() => {
    if (props.color) {
        if (is_filled.value) {
            return getTextColorForBackground(props.color);
        } else {
            return props.color;
        }
    }

    const colors = is_filled.value ? filled_circle_variant_colors : design_colors;
    return colors[props.design] || null;
});

const design_colors: Record<`${IkUIDesignColor}`, string | null> = {
    primary: theme.colors.value.primary_color,
    accent: theme.colors.value.accent_color,
    error: theme.colors.value.error_color,
    success: theme.colors.value.success_color,
    default: theme.colors.value.on_surface_color_3,
};

const filled_circle_variant_colors: Record<`${IkUIDesignColor}`, string | null> = {
    primary: theme.colors.value.on_primary_color,
    accent: theme.colors.value.on_accent_color,
    error: theme.colors.value.on_error_color,
    success: theme.colors.value.on_success_color,
    default: theme.colors.value.on_surface_color_3,
};

function onLoad(svg_string: string) {
    renderSvg(svg_string);
}

function renderSvg(svg_string: string) {
    if (icon.value) {
        icon.value.innerHTML = '';
        icon.value.insertAdjacentHTML('afterbegin', svg_string);

        const nodes = icon.value.firstElementChild ? getNodes(icon.value.firstElementChild) : [];

        nodes.forEach(node => {
            if (
                node.hasAttribute('fill') &&
                node.getAttribute('fill') !== '' &&
                node.getAttribute('fill') !== 'none' &&
                stroke.value
            ) {
                node.setAttribute('fill', `#${stroke.value}`);
            }
            if (
                node.hasAttribute('stroke') &&
                node.getAttribute('stroke') !== '' &&
                node.getAttribute('stroke') !== 'none' &&
                stroke.value
            ) {
                node.setAttribute('stroke', `#${stroke.value}`);
            }
        });
        const svg = icon.value.firstElementChild;

        if (props.circle || is_filled.value) {
            svg?.setAttribute('width', (100 * clamp(props.inner_scale, 0, 1)) + '%');
            svg?.setAttribute('height', (100 * clamp(props.inner_scale, 0, 1)) + '%');
        } else {
            svg?.setAttribute('width', '100%');
            svg?.setAttribute('height', '100%');
        }

        if (props.circle) {
            svg?.setAttribute('aspect-ratio', '1');
        }
    }
}

function getNodes(node: Element): Element[] {
    const nodes: Element[] = [];
    if (node.nodeType === Node.ELEMENT_NODE) {
        nodes.push(node);
        const len = node.children.length;
        for (let i = 0; i < len; i++) {
            nodes.push(...getNodes(node.children[i]));
        }
    }
    return nodes;
}

function load() {
    if (props.src) {
        fetchSvgText(props.src)
            .then(onLoad);
    }
}

onMounted(() => {
    load();
});

watch(() => props.src, () => {
    load();
});

watch(() => props.circle, () => {
    load();
});

watch(is_filled, () => {
    load();
});

watch(stroke, () => {
    load();
});

useRender(() => {
    return h('div', {
        class: {
            'ik-svg-icon': true,
            'ik-svg-icon--circle': props.circle,
            'ik-svg-icon--clickable': !props.disabled && !!attrs.onClick,
            'ik-svg-icon--disabled': props.disabled,
            ['ik-svg-icon--' + props.design]: !!props.design,
            ['ik-svg-icon--' + props.variant]: !!props.variant,
        },
        ref: icon,
        style: {
            width: toCssUnits(props.size_px, 'px'),
            height: toCssUnits(props.size_px, 'px'),
            backgroundColor: bg_color.value,
        },
    });
});
</script>
