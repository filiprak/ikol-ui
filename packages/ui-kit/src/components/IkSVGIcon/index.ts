import IkSVGIcon from './IkSVGIcon.vue';
export type IkSVGIconT = InstanceType<typeof IkSVGIcon>;
export { IkSVGIcon };
