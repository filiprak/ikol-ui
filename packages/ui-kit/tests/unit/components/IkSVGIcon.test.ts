import { renderShallow } from '@ui/tests/tools/utils';
import { IkSVGIcon } from '@ui/components/IkSVGIcon';

describe('IkSVGIcon', () => {
    const getInstance = (options = {}) => renderShallow(IkSVGIcon, options);

    it('renders with correct default state', () => {
        const wrapper = getInstance();
        expect(wrapper.classes()).toContain('ik-svg-icon');
        expect(wrapper.element.tagName).toBe('DIV');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with correct src', () => {
        const src = '/docs/stories/assets/plugin.svg';

        getInstance({
            props: {
                src,
            },
        });

        expect(window.fetch).toHaveBeenCalledWith(src);
    });

    it('renders as circle', () => {
        const wrapper = getInstance({
            props: {
                circle: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-svg-icon--circle');
        expect(wrapper.attributes().style).toContain('width: 28px');
        expect(wrapper.attributes().style).toContain('height: 28px');
    });

    it('renders with pixel size', () => {
        const wrapper = getInstance({
            props: {
                size_px: 35,
            },
        });

        expect(Object.keys(wrapper.attributes())).toContain('style');
        expect(wrapper.attributes().style).toContain('width: 35px');
        expect(wrapper.attributes().style).toContain('height: 35px');
    });

    it('renders with pixel size as circle', () => {
        const wrapper = getInstance({
            props: {
                size_px: 35,
                circle: true,
            },
        });

        expect(Object.keys(wrapper.attributes())).toContain('style');
        expect(wrapper.classes()).toContain('ik-svg-icon--circle');
        expect(wrapper.attributes().style).toContain('width: 35px');
        expect(wrapper.attributes().style).toContain('height: 35px');
    });

    it('renders with design', () => {
        const wrapper = getInstance({
            props: {
                design: 'success',
            },
        });

        expect(wrapper.classes()).toContain('ik-svg-icon--success');
    });

    it('renders clickable', async () => {
        const clickFn = jest.fn();
        const wrapper = getInstance({
            props: {
                onClick: clickFn,
            },
        });

        await wrapper.find('.ik-svg-icon').trigger('click');

        expect(clickFn).toHaveBeenCalled();
        expect(wrapper.classes()).toContain('ik-svg-icon--clickable');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders disabled', () => {
        const wrapper = getInstance({
            props: {
                disabled: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-svg-icon--disabled');
        expect(wrapper.classes()).not.toContain('ik-svg-icon--clickable');
        expect(wrapper.element).toMatchSnapshot();
    });
});
