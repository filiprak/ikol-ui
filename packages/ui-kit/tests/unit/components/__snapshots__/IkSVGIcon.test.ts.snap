// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkSVGIcon renders clickable 1`] = `
<div
  class="ik-svg-icon ik-svg-icon--clickable ik-svg-icon--default"
  style="width: 28px; height: 28px;"
/>
`;

exports[`IkSVGIcon renders disabled 1`] = `
<div
  class="ik-svg-icon ik-svg-icon--disabled ik-svg-icon--default"
  style="width: 28px; height: 28px;"
/>
`;

exports[`IkSVGIcon renders with correct default state 1`] = `
<div
  class="ik-svg-icon ik-svg-icon--default"
  style="width: 28px; height: 28px;"
/>
`;
